lua-mosquitto.c

```diff
@@ -432,6 +432,56 @@ static int ctx_tls_set(lua_State *L)
 	return mosq__pstatus(L, rc);
 }
 
+struct mosq_opt_string {
+	const char *str;
+	enum mosq_opt_t tls_opt;
+};
+
+/***
+ * Enables the use of a TLS engine device.
+ * @function string_option
+ * @tparam[opt=nil] string option may be nil
+ * @tparam[opt=nil] string value may be nil
+ * @see mosquitto_string_option
+ * @return[1] boolean true
+ * @return[2] nil
+ * @treturn[2] number error code
+ * @treturn[2] string error description.
+ * @raise For some out of memory or illegal states
+ */
+static int ctx_string_option(lua_State *L)
+{
+       ctx_t *ctx = ctx_check(L, 1);
+       const char *mosq_option = luaL_optstring(L, 2, NULL);
+       const char *value = luaL_optstring(L, 3, NULL);
+       enum mosq_opt_t option;
+
+       struct mosq_opt_string options[] = {
+	       {"MOSQ_OPT_PROTOCOL_VERSION", MOSQ_OPT_PROTOCOL_VERSION},
+	       {"MOSQ_OPT_SSL_CTX", MOSQ_OPT_SSL_CTX},
+	       {"MOSQ_OPT_SSL_CTX_WITH_DEFAULTS", MOSQ_OPT_SSL_CTX_WITH_DEFAULTS},
+	       {"MOSQ_OPT_RECEIVE_MAXIMUM", MOSQ_OPT_RECEIVE_MAXIMUM},
+	       {"MOSQ_OPT_SEND_MAXIMUM", MOSQ_OPT_SEND_MAXIMUM},
+	       {"MOSQ_OPT_TLS_KEYFORM", MOSQ_OPT_TLS_KEYFORM},
+	       {"MOSQ_OPT_TLS_ENGINE", MOSQ_OPT_TLS_ENGINE},
+	       {"MOSQ_OPT_TLS_ENGINE_KPASS_SHA1", MOSQ_OPT_TLS_ENGINE_KPASS_SHA1},
+	       {"MOSQ_OPT_TLS_OCSP_REQUIRED", MOSQ_OPT_TLS_OCSP_REQUIRED},
+	       {"MOSQ_OPT_TLS_ALPN", MOSQ_OPT_TLS_ALPN}
+       };
```

Comment:
You probably want to provide these symbolicly to the user, rather than requiring them to pass in strings?

.option(lmosq.OPT_SSL_CTX, "blah") is nicer than .option("MOSQ_OPT_SSL_CTX", "blah") though they should probably both work

Reply:
You can just add them all to the list at https://github.com/flukso/lua-mosquitto/blob/master/lua-mosquitto.c#L1082-L1100 and they'll be automatically available for both styles, and you can then just not even worry abotu checking, just attempt to call the mosquitto function _string_option or _int_option based on the type of the argument and handle the results from the lib